### PR TITLE
Initial linking tool

### PR DESCRIPTION
**Pull request by ManInFez**
Resolves https://github.com/equinor/scout/issues/110

### CODE REVIEW

File: tests/test_link_io_structure.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
+from symlink_komodo.sanity_check import read_link_structure, compare_links, check_root_nodes, verify_integrity
+from symlink_komodo.create_links import create_symlinks
+
+import os
+from shutil import copy
+import json
+
+
+def test_read_folder_structure(tmpdir):
+    with tmpdir.as_cwd() as old_dir:
```
Comment (markusdregi) [inline tests/test_link_io_structure.py]: Would not `with tmpdir.as_cwd()` suffice?

File: tests/test_link_io_structure.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        assert compare_links(input_dict, output)
+
+
+def test_root_links(tmpdir):
+    
+    test_file = os.getenv('PYTEST_CURRENT_TEST')
+    test_folder = os.path.dirname(test_file)
+    copy(os.path.join(test_folder, 'data/links.yml'), tmpdir)
+    with tmpdir.as_cwd() as old_dir:
+        os.chdir(tmpdir)
```
Comment (markusdregi) [inline tests/test_link_io_structure.py]: Is this necessary?

File: tests/test_link_io_structure.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
+from symlink_komodo.sanity_check import read_link_structure, compare_links, check_root_nodes, verify_integrity
+from symlink_komodo.create_links import create_symlinks
+
+import os
+from shutil import copy
```
Comment (markusdregi) [inline tests/test_link_io_structure.py]: No reason to drop the entire namespace here. I think `import shutil` followed by calling `shutil.copy` is a better solution...

File: tests/test_link_io_structure.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        assert os.path.realpath('unstable') == os.path.realpath('2012.01.rc2')
+
+
+def test_integration(tmpdir):
+    
+    test_file = os.getenv('PYTEST_CURRENT_TEST')
+    test_folder = os.path.dirname(test_file)
+    copy(os.path.join(test_folder, 'data/links.yml'), tmpdir)
+    with tmpdir.as_cwd() as old_dir:
+        input_dict = json.load(open('links.yml'))
```
Comment (markusdregi) [inline tests/test_link_io_structure.py]: `json.loads("links.yml")` instead

File: tests/test_link_io_structure.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+def test_root_links(tmpdir):
+    
+    test_file = os.getenv('PYTEST_CURRENT_TEST')
+    test_folder = os.path.dirname(test_file)
+    copy(os.path.join(test_folder, 'data/links.yml'), tmpdir)
+    with tmpdir.as_cwd() as old_dir:
+        os.chdir(tmpdir)
+        input_dict = json.load(open('links.yml'))
+        
+        check_root_nodes(input_dict)
```
Comment (markusdregi) [inline tests/test_link_io_structure.py]: How about naming this one `assert_root_nodes` and then raising an `AssertionError` instead?

File: tests/test_link_io_structure.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+        assert os.path.exists('stable')
+        assert os.path.islink('stable')
+        assert os.readlink('stable') == '2012.01'
+        assert os.path.realpath('stable') == os.path.realpath('2012.01.12')
+        assert os.path.realpath('unstable') == os.path.realpath('2012.01.rc2')
+
+
+def test_integration(tmpdir):
+    
+    test_file = os.getenv('PYTEST_CURRENT_TEST')
```
Comment (markusdregi) [inline tests/test_link_io_structure.py]: Could you please explain this?

File: tests/data/links.yml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  ],
+  "links": {
+    "2012.01": "2012.01.12",
+    "unstable": "2012.01.rc1",
+    "testing": "2012.01",
+    "stable" : "testing",
+    "bleeding": "testing"
+  }
+}
+
```
[the remark this answers is not in the document]
Comment (ManInFez) [reply tests/data/links.yml]: I feel it averages out quite nicely...?

File: tests/data/links_full.yml [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    "testing"       : "testing-py3",
+  
+    "stable-py36"   : "2019.12.00-py36", 
+    "stable-py3"    : "stable-py36",
+    "stable-py27"   : "2019.12.00-py27",
+    "stable-py2"    : "stable-py27",
+    "stable"        : "stable-py3"
+
+  }
+}
```
[the remark this answers is not in the document]
Comment (ManInFez) [reply tests/data/links_full.yml]: :+1:

File: symlink_komodo/sanity_check.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
+import os, sys
+import json
+import argparse
+import difflib
+import pprint
+
+
+def compare_links(a, b):
```
[the remark this answers is not in the document]
Comment (ManInFez) [reply symlink_komodo/sanity_check.py]: :+1:

File: tests/test_link_io_structure.py (hunk 4 above)
Comment (ManInFez) [reply tests/test_link_io_structure.py]: json.loads expects a json as a string, not a filename

File: tests/test_link_io_structure.py (hunk 3 above)
Comment (ManInFez) [reply tests/test_link_io_structure.py]: :+1:

File: tests/test_link_io_structure.py (hunk 5 above)
Comment (ManInFez) [reply tests/test_link_io_structure.py]: :+1:

File: tests/test_link_io_structure.py (hunk 2 above)
Comment (ManInFez) [reply tests/test_link_io_structure.py]: :+1:

File: tests/test_link_io_structure.py (hunk 1 above)
Comment (ManInFez) [reply tests/test_link_io_structure.py]: :+1:

File: komodo/symlink_komodo/sanity_check.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        raise AssertionError("The roots in the link-tree is not matching the roots defined in link_roots in dict")
+
+
+def _compare_dicts(d1, d2):
+    return ('\n' + '\n'.join(difflib.ndiff(
+                   pprint.pformat(d1).splitlines(),
+                   pprint.pformat(d2).splitlines())))
+
+def sanity_main():
+    parser = argparse.ArgumentParser(description='Create symlinks for komodo versions.')
```
Comment (markusdregi) [inline komodo/symlink_komodo/sanity_check.py]: This one does not create symlinks

Comment (ManInFez) [reply komodo/symlink_komodo/sanity_check.py]: :+1:

File: komodo/symlink_komodo/create_links.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+def symlink_main():
+
+    parser = argparse.ArgumentParser(description='Create symlinks for komodo versions.')
+    parser.add_argument('config', type=str,
+                        help='a json file describing symlink structure')
+
+    args = parser.parse_args()
+    if not os.path.isfile(args.config):
+        print('the file {} can not be found'.format(args.config))
```
[the remark this answers is not in the document]
Comment (ManInFez) [reply komodo/symlink_komodo/create_links.py]: :+1:

File: komodo/symlink_komodo/create_links.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+        sys.exit(1)
+    
+    input_dict = json.load(open(args.config))
```
[the remark this answers is not in the document]
Comment (ManInFez) [reply komodo/symlink_komodo/create_links.py]: :+1:

File: komodo/symlink_komodo/create_links.py [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    
+    os.symlink(src, dst)
+
+
+def create_symlinks(links_dict):
+    root_folder = links_dict['root_folder']
+    if not os.path.isabs(root_folder):
+        raise ValueError("The root folder specified is not absolute")
+
+    if not os.path.exists(root_folder):
```
[the remark this answers is not in the document]
Comment (ManInFez) [reply komodo/symlink_komodo/create_links.py]: :+1:

File: komodo/symlink_komodo/create_links.py [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        raise ValueError("The folder {} does not exist".format(root_folder))
+
+    os.chdir(root_folder)
```
[the remark this answers is not in the document]
Comment (ManInFez) [reply komodo/symlink_komodo/create_links.py]: :+1:

File: tests/test_link_io_structure.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
+from komodo.symlink_komodo.sanity_check import read_link_structure, equal_links, assert_root_nodes, verify_integrity
```
[the remark this answers is not in the document]
Comment (ManInFez) [reply tests/test_link_io_structure.py]: I disagree, as this clutters the code when the namespace is too deep

[the remark this answers is not in the document]
Comment (markusdregi) [reply tests/test_link_io_structure.py]: My counter argument is that then your namespace structure is wrong ;)

[the remark this answers is not in the document]
Comment (markusdregi) [reply tests/test_link_io_structure.py]: It is of course not that black and white, but I think that it should be able to make an API for the komodo package that is natural and easy to use based solely on importing the komodo module.

File: komodo/symlink/create_links.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    finally:
+        os.chdir(prev_dir)
+
+def _create_link(src, dst, link_dict):  
+    if src in link_dict and not os.path.exists(src):
+        _create_link(link_dict[src], src, link_dict)
+
+    if not os.path.exists(src):
+        raise ValueError("{} does not exist".format(src))
+
```
Comment (markusdregi) [inline komodo/symlink/create_links.py]: To minimise the number of touches on the disk and the chances for syncing something broken, should we do something like:

```python
if os.path.exists(dst) and os.path.islink(dst):
    if os.readlink(dst) == src:
        return
    else:
        os.remove(dst)
```

File: komodo/symlink/create_links.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    if not os.path.isabs(root_folder):
+        raise ValueError("The root folder specified is not absolute")
+
+    if not os.path.isdir(root_folder):
+        raise ValueError("The folder {} does not exist".format(root_folder))
+
+    with working_dir(root_folder):  
+        for dst, src in links_dict['links'].items():
+            _create_link(src, dst, links_dict['links'])
+
```
Comment (markusdregi) [inline komodo/symlink/create_links.py]: Should be double line...

File: komodo/symlink/sanity_check.py [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+        return False
+
+    if set(a['root_links']) != set(b['root_links']):
+        return False
+
+    return a['links'] == b['links']
+
+
+def _linked_to(s_link, list_of_files):
+
```
Comment (markusdregi) [inline komodo/symlink/sanity_check.py]: Remove empty line

File: komodo/symlink/sanity_check.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+        file_name = os.path.basename(file_path)
+        if os.path.islink(file_path):
+            link_structure['links'][file_name] = os.path.basename(os.readlink(file_path))
+            if not _linked_to(file_name, list_of_files):
+                link_structure['root_links'].append(file_name)
+
+    return link_structure
+
+
+def _check_link(link, link_dict, errors, visited=[]):
```
Comment (markusdregi) [inline komodo/symlink/sanity_check.py]: Having mutable objects as default values are super scary. The recommended practice is something like:
```python
def (...., visited=None):
    if visited is None:
        visited = []
```

or to use an immutable structure:
```python
def (...., visited=()):
    # And then either:
    visited = list(visited)
    # or when adding
    visited += (link,)
```

File: komodo/symlink/sanity_check.py [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+    visited.append(link)
+
+    if link in link_dict['links'].keys():
+        _check_link(link_dict['links'][link], link_dict, errors, visited)
+
+    elif not os.path.exists(os.path.join(link_dict['root_folder'], link)):
+        error = "{} does not exist".format(link)
+        if error not in errors:
+            errors.append(error)
+
```
Comment (markusdregi) [inline komodo/symlink/sanity_check.py]: Missing new line

File: tests/test_link_io_structure.py [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+                'unstable': '2012.01.rc2'
+            }
+        }
+
+        output_dict = read_link_structure(os.getcwd())
+
+        assert type(output_dict) == dict
+        assert equal_links(output_dict, expected_result)
+
+
```
Comment (markusdregi) [inline tests/test_link_io_structure.py]: Please ensure 2 new lines between functions in this file ;)

File: tests/test_link_io_structure.py [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+            os.mkdir('2020.01.a0-py36')
+            os.mkdir('2020.01.a0-py27')
+            os.mkdir('2019.12.00-py36')
+            os.mkdir('2019.12.00-py27')
+            
+            symlink_main()
+            
+            sanity_main()
+    finally:
+        sys.argv = old_argv
```
Comment (markusdregi) [inline tests/test_link_io_structure.py]: Missing new line

File: komodo/symlink/create_links.py (hunk 16 above)
Comment (ManInFez) [reply komodo/symlink/create_links.py]: :+1:

File: komodo/symlink/sanity_check.py (hunk 20 above)
Comment (ManInFez) [reply komodo/symlink/sanity_check.py]: :+1:

File: tests/test_link_io_structure.py (hunk 21 above)
Comment (ManInFez) [reply tests/test_link_io_structure.py]: :+1:

File: komodo/symlink/sanity_check.py (hunk 19 above)
Comment (ManInFez) [reply komodo/symlink/sanity_check.py]: I think not having a default value is probably just as well. Was not aware of this issue..

File: komodo/symlink/create_links.py (hunk 17 above)
Comment (ManInFez) [reply komodo/symlink/create_links.py]: :+1:

**Pull request merged**